crates/rethnet_evm/src/state/fork.rs

```diff
@@ -1,82 +1,112 @@
-    /// An error from the underlying LayeredDatabase
-    #[error(transparent)]
-    LayeredDatabase(#[from] <LayeredDatabase<RethnetLayer> as revm::Database>::Error),
-
-    /// Code hash not found in cache of remote database
-    #[error("Cache of remote database does not contain contract with code hash: {0}.")]
-    NoSuchCodeHash(B256),
-
-    /// Some other error from an underlying dependency
-    #[error(transparent)]
-    OtherError(#[from] std::io::Error),
+    fork_block_number: U256,
+    /// client-facing state root (pseudorandomly generated) mapped to internal (layered_state) state root
+    state_root_to_state: HashMap<B256, B256>,
+    /// for mapping historical blocks to their state roots
+    state_root_to_block_number: HashMap<B256, U256>,
+    latest_state_root: B256,
 }
 
-impl ForkDatabase {
-    /// instantiate a new ForkDatabase
-    pub fn new(url: &str, fork_block_number: u64) -> Self {
-        let remote_db = RemoteDatabase::new(url);
+use crate::random::RandomHashGenerator;
+use once_cell::sync::Lazy;
+use std::sync::Mutex;
+static RANDOM_HASH_GENERATOR: Lazy<Mutex<RandomHashGenerator>> =
+    Lazy::new(|| Mutex::new(RandomHashGenerator::with_seed("seed")));
+
+impl ForkState {
+    /// instantiate a new ForkState
+    pub fn new(
+        url: &str,
+        accounts: HashMap<Address, AccountInfo>,
+        fork_block_number: Option<U256>,
+    ) -> Self {
+        let rpc_client = RpcClient::new(url);
+
+        let async_runtime = Builder::new_multi_thread()
+            .enable_io()
+            .enable_time()
+            .build()
+            .expect("failed to construct async runtime");
+
+        let fork_block_number = fork_block_number
+            .or_else(|| {
+                async_runtime
+                    .block_on(rpc_client.get_latest_block())
+                    .expect("failed to get latest block")
+                    .number
+                    .map(U256::from)
+            })
+            .unwrap();
+
+        let remote_state = RemoteState::new(url, fork_block_number);
+
+        let mut initialized_accounts = accounts.clone();
+        for (address, account_info) in accounts.iter() {
+            let remote_account_info = async_runtime
+                .block_on(
+                    rpc_client.get_account_info(address, BlockSpec::Number(fork_block_number)),
+                )
+                .expect("failed to retrieve remote account info for local account initialization");
+            initialized_accounts.insert(
+                *address,
+                AccountInfo {
+                    nonce: remote_account_info.nonce,
+                    ..account_info.clone()
+                },
+            );
+        }
 
-        let layered_db = LayeredDatabase::<RethnetLayer>::default();
+        let mut layered_state =
+            LayeredState::with_layer(RethnetLayer::with_genesis_accounts(initialized_accounts));
+        crate::state::StateDebug::checkpoint(&mut layered_state).unwrap();
```

Comment:
This introduces a discrepancy with how the `LayeredState` works. It assumes that the starting layer is initialized with the genesis accounts and no checkpoint is created. The `napi` wrapper calls checkpoint to ensure that genesis block is always checkpointed. We should do the same in both places.

In the case of the `LayeredState` it's documented. `with_layer` creates the database with the provided layer as its only layer - ergo no checkpoint.